V4

kyeotic opened the pull request:
closes #113
closes #115
closes #116 
closes #117

This PR replaces #114

I would like to have a codemod for the change to `navigate`, but if I am being realistic I don't have the energy to put one together. I just started a new job and will be moving out of state in a few months, so my OSS time is extremely limited. If someone wants to write one I would be very grateful. 

**Changelog Preview**
---
## [4.0.0] - Unreleased
### Added
- `useHistory`
- `usePath`
### Changed
- **BREAKING**: `navigate` now has `options` object instead of overloads
- **BREAKING**: `useNavigate` uses updated `navigate` params
- **BREAKING**: `useLocationChange` now invokes the setter with a `RavigerLocation` object instead of the `path` string
### Fixed
-  `usePath` getting an old path if `navigate` was called during the initial render
---

~The docs have not yet been updated for any of this. **External Contributions Welcome**~
Edit: docs have been updated

I have released a beta for this code, it can be installed with

```
npm i raviger@beta
// OR
npm i raviger@4.0.0-2
```

dave-irvine commented:
Hmm, something still isn't working right. Here's my earlier example but updated to the v4 release:

https://codesandbox.io/s/crazy-spence-qpsr4

Navigating between screens works fine initially. Then if you reload the app while on the `/error` path, this clears your history state, so the app navigates you back to `/`. Now clicking the buttons does nothing.

https://user-images.githubusercontent.com/1782643/153190143-8feda33c-12ac-4fd0-a6d6-917ffe0fbbfa.mp4


Sidenote: If I wanted to add the docs for this release, whats the best way of doing that? Forking the main repo, checking out the v4 branch and adding the docs there, then raising a pull request against *your* v4 branch?

I'd love to help with a codemod but I can't honestly say I understand them well enough to help you out here.

timkye-blueocean-ai commented:
> Sidenote: If I wanted to add the docs for this release, whats the best way of doing that? Forking the main repo, checking out the v4 branch and adding the docs there, then raising a pull request against your v4 branch?

Yes, that would work.

I will dig into the broken button issue, thanks for the gif 😄

kyeotic commented:
@dave-irvine Ive updated the beta (4.0.0-1) to include the `useHistory` export

kyeotic commented:
@dave-irvine with the latest beta [the sandbox](https://codesandbox.io/s/fervent-cache-11zp6g) appears to work correctly. Hopefully you can verify this is what you expect to see.

Today I will be updating the docs and creating a v4 migration guide. Unless any issues come up I will release to latest sometime this week.

kyeotic commented:
@dave-irvine I've released a new beta that removes the `usePathChange` function, but I've updated the docs and added a migration guide that includes the method for those that want it

```
npm i raviger@beta
// OR
npm i raviger@4.0.0-2
```

dave-irvine commented:
Hmm, appears to not be working in the sandbox as per the screen recording. Here's the latest sandbox version pointing at 4.0.0-2 https://codesandbox.io/s/delicate-monad-y3ts5c

Once it loads, navigate to the Error route by clicking the "Go to error" button. Now click the "refresh" button in the codesandbox UI (not your actual browser refresh). This will reset the path to "/" and render the Main page. Now click the "Go to error" button again. The path changes to "/error", but the page does not re-render to show the error screen.

kyeotic commented:
Ok, I missed which button was breaking. The repro helps, I will sort it out

dave-irvine commented:
Yeah great it seems like this works! I'm super confused about what you actually changed in your commits, but the end result is fine :)

timkye-blueocean-ai commented:
@dave-irvine [This](https://github.com/kyeotic/raviger/pull/118/commits/7a37735255323ad0a7894b9847657d98ea3c5188#diff-563a75de07924a14c7aa91020a763d60ff7682b796c2033d7e3ae9b1cdf704cdR57-R58) is what ended up fixing it.
